src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -147,31 +98,57 @@ private SocketError ProcessIOCPResult(bool success, int bytesTransferred)
                     FinishOperationSyncSuccess(bytesTransferred, SocketFlags.None);
                     return SocketError.Success;
                 }
-
-                // Socket handle is going to post a completion to the completion port (may have done so already).
-                // Return pending and we will continue in the completion port callback.
-                return SocketError.IOPending;
             }
-
-            // Get the socket error (which may be IOPending)
-            SocketError errorCode = SocketPal.GetLastSocketError();
-
-            if (errorCode == SocketError.IOPending)
+            else
             {
-                return errorCode;
+                // Get the socket error (which may be IOPending)
+                SocketError socketError = SocketPal.GetLastSocketError();
+                if (socketError != SocketError.IOPending)
+                {
+                    FinishOperationSyncFailure(socketError, bytesTransferred, SocketFlags.None);
+                    return socketError;
+                }
             }
 
-            FinishOperationSyncFailure(errorCode, bytesTransferred, SocketFlags.None);
-
-            return errorCode;
+            // Socket handle is going to post a completion to the completion port (may have done so already).
+            // Return pending and we will continue in the completion port callback.
+            return SocketError.IOPending;
         }
 
-        private void InnerStartOperationAccept(bool userSuppliedBuffer)
+        private SocketError ProcessIOCPResultWithSingleBufferHandle(SocketError socketError, int bytesTransferred)
         {
-            if (!userSuppliedBuffer)
+            if (socketError == SocketError.Success)
             {
-                CheckPinSingleBuffer(false);
+                // Synchronous success.
+                if (_currentSocket.SafeHandle.SkipCompletionPortOnSuccess)
+                {
+                    // The socket handle is configured to skip completion on success, 
+                    // so we can set the results right now.
+                    _singleBufferHandleState = SingleBufferHandleState.None;
```

Comment:
I assume that at this point, we're expecting _singleBufferHandleState to be InProcess, is that correct?  Can we assert this?

Reply:
It could be InProcess, or it could be Set if the operation did Retain prior to initiating the call.

I don't love this handling of it, but I couldn't come up with anything better.

Reply:
Yeah, I understand how it works now.  Seems reasonable.